graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/BuiltinFunctions.java

```diff
@@ -286,6 +296,154 @@ public Object absObject(VirtualFrame frame, Object object,
         }
     }
 
+    @Builtin(name = ALL, minNumOfPositionalArgs = 1)
+    @GenerateNodeFactory
+    public abstract static class AllNode extends PythonUnaryBuiltinNode {
+        @Specialization
+        static boolean doList(VirtualFrame frame,
+                             PList list,
+                             @Cached PyObjectIsTrueNode isTrueNode) {
+            return checkSequenceStorage(list, frame, isTrueNode);
+        }
+
+        @Specialization
+        static boolean doTuple(VirtualFrame frame,
+                               PTuple tuple,
+                               @Cached PyObjectIsTrueNode isTrueNode) {
+            return checkSequenceStorage(tuple, frame, isTrueNode);
+        }
+
+        @Specialization
+        static boolean doDict(VirtualFrame frame,
+                              PDict hc,
+                              @Cached PyObjectIsTrueNode isTrueNode) {
+            return true;
+        }
+
+        @Specialization
+        static boolean doSet(VirtualFrame frame,
+                             PSet hc,
+                             @Cached PyObjectIsTrueNode isTrueNode) {
+            return checkHashEntries(hc, frame, isTrueNode);
+        }
+
+        @Specialization
+        static boolean doObject(VirtualFrame frame,
+                                Object object,
+                                @Cached PyObjectGetIter getIter,
+                                @Cached("create(__NEXT__)") LookupAndCallUnaryNode callNode,
+                                @Cached IsBuiltinClassProfile errorProfile,
+                                @Cached PyObjectIsTrueNode isTrueNode) {
+            Object iterator = getIter.execute(frame, object);
+            while (true) {
+                try {
+                    Object next = callNode.executeObject(frame, iterator);
+                    if (!isTrueNode.execute(frame, next))
+                        return false;
+                } catch (PException e) {
+                    e.expectStopIteration(errorProfile);
+                    break;
+                }
+            }
+
+            return true;
+        }
+
+        static boolean checkSequenceStorage(PSequence seq,
+                                            VirtualFrame frame,
+                                            PyObjectIsTrueNode isTrueNode) {
+            SequenceStorage sequenceStorage = seq.getSequenceStorage();
+            Object[] internalArray = sequenceStorage.getInternalArray();
+            for (int i = 0; i < sequenceStorage.length(); i++)
```

Comment:
Calling virtual methods like `sequenceStorage.length()` that have multiple implementations cannot always get inlined which has its performance cost. Under Truffle, we always try to avoid such virtual calls by making sure that the compiler has enough information to pick just one implementation, mostly by specializing on leaf types. For example, a few lines back you call `seq.getSequenceStorage()` and that is fine because this method will get inlined into the list (and tuple) specialization where the compiler knows that `seq`  is `PList` (`PTuple`) and the method in `PList` is `final`. But there's no such type information for the `length()` call here, the `sequenceStorage` could be any of the `SequenceStorages` subtypes.
For sequence storages, we have nodes that will try to specialize on the types such that virtual calls are avoided in most cases. For `length()`, there is `com.oracle.graal.python.builtins.objects.common.SequenceStorageNodes.LenNode`.
If you specialize on the storages as I outlined in the comment above, you can just call `length()` in specializations that specialize on a leaf type (i.e. `BooleanSequenceStorage`), but you will need `LenNode` if you have a generic fallback that covers more storages at the same time.